File: BaekJoon/Solutions/9012_괄호/solution.dart
import 'dart:io';

void forLoop(int limit, Function(int index) fn) {
  for (int i = 0; i < limit; i++) {
    fn(i);
  }
}

void main() {
  String? TStr = stdin.readLineSync();

  if (TStr == null) {
    return;
  }

  int T = int.parse(TStr);

  forLoop(T, (index) {
    String? testCase = stdin.readLineSync();
    if (testCase != null) {
      List<String> stack = [];

      forLoop(testCase.length, (index) {
        String parenthesis = testCase[index];
        if (parenthesis == '(') {
          stack.add('(');
        } else {
          if (stack.isNotEmpty && stack.last == '(') {
            stack.removeLast();
          } else {
            stack.add(')');
          }
        }
      });

      print(stack.length == 0 ? 'YES' : 'NO');
    }
  });
}

File: BaekJoon/Solutions/9012_괄호/solution.cpp
#include <iostream>
#include <stdlib.h>
#include <vector>

using namespace std;

int main()
{
    int T;
    cin >> T;

    for (int i = 0; i < T; i++)
    {
        string testCase;
        cin >> testCase;
        vector<char> stack = {};

        for (int j = 0; j < testCase.length(); j++)
        {
            if (testCase[j] == '(')
            {
                stack.push_back('(');
            }
            else
            {
                if (stack.size() != 0 && stack.back() == '(')
                {
                    stack.pop_back();
                }
                else
                {
                    stack.push_back('0');
                }
            }
        }

        cout << (stack.empty() ? "YES" : "NO") << endl;
    }

    return 0;
}